pkg/collector/container_cgroup_collector.go

```diff
@@ -25,15 +25,34 @@ import (
 
 // updateCgroupMetrics adds container-level cgroup data
 func (c *Collector) updateCgroupMetrics() {
-	klog.V(5).Infof("overall cgroup stats %v", cgroup.SliceHandlerInstance)
-
-	for containerID := range c.ContainersMetrics {
-		cgroup.TryInitStatReaders(containerID)
-		cgroupFSStandardStats := cgroup.GetStandardStat(containerID)
-		for cgroupFSKey, cgroupFSValue := range cgroupFSStandardStats {
-			readVal := cgroupFSValue.(uint64)
-			if _, ok := c.ContainersMetrics[containerID].CgroupFSStats[cgroupFSKey]; ok {
-				c.ContainersMetrics[containerID].CgroupFSStats[cgroupFSKey].SetAggrStat(containerID, readVal)
+	for key := range c.ContainersMetrics {
+		if c.ContainersMetrics[key].PID == 0 {
+			klog.V(3).Infof("PID 0 does not have cgroup")
+			return
+		}
+		if c.ContainersMetrics[key].CgroupStatHandler == nil {
+			handler, err := cgroup.NewCGroupStatHandler(int(c.ContainersMetrics[key].PID))
+			if err != nil {
+				klog.V(3).Infoln("Error: could not start cgroup stat handler for PID:", c.ContainersMetrics[key].PID)
+				return
+			}
+			c.ContainersMetrics[key].CgroupStatHandler = handler
+		}
+
+		// we need to check again if CgroupStatHandler is not nil because on darwin OS the handler will always be nil
+		if c.ContainersMetrics[key].CgroupStatHandler != nil {
```

Comment:
maybe we can have another PR to update CgroupStatMap from GetCGroupStat without cGroupMetrics as a by pass value to reduce memory alloc and GC.